zerver/actions/message_delete.py

```diff
@@ -52,6 +52,28 @@ def do_delete_messages(realm: Realm, messages: Iterable[Message]) -> None:
         archiving_chunk_size = retention.STREAM_MESSAGE_BATCH_SIZE
 
     move_messages_to_archive(message_ids, realm=realm, chunk_size=archiving_chunk_size)
+    if message_type == "stream": # nocoverage
+        stream = Stream.objects.get(id=sample_message.recipient.type_id)
+        if stream.first_message_id:
+            first_stream_message_exists = Message.objects.filter(
+                id=stream.first_message_id
+            ).exists()
+            if not first_stream_message_exists and stream.recipient_id:
+                stream_messages = Message.objects.filter(
+                    realm_id=realm.id, recipient_id=stream.recipient_id
+                ).order_by("date_sent")
+                if stream_messages:
+                    stream.first_message_id = stream_messages[0].id
+                    stream.save(update_fields=["first_message_id"])
+
+                    stream_event = dict(
+                        type="stream",
+                        op="update",
+                        property="first_message_id",
+                        value=stream.first_message_id,
+                        stream_id=stream.id,
+                    )
+                    send_event(realm, stream_event, users_to_notify)
```

Comment:
Basically any backend change like this should include tests in the first PR for it -- it's easy to write good tests that verify logic, and writing them helps a great deal with writing the code well -- since it forces you to think about corner cases. And the work is rarely wasted, since if you're writing a good end-to-end test, the test should still be correct even if you replace the implementation completely.

See also https://zulip.readthedocs.io/en/latest/testing/philosophy.html.

We don't use `nocoverage` comments in the server without a very good reason.